### PR TITLE
JSON schema inference tool.

### PR DESCRIPTION
**Pull request by dkorolev**
Hi @mzhurovich ,

Got to the pull request stage. Could you please take a look?

Thanks,
Dima

### CODE REVIEW

File: Utils/JSONSchema/infer.h [hunk 1]
```diff
@@ -0,0 +1,542 @@
...
+inline std::string MaybeOptionalHumanReadableType(bool has_nulls, const std::string& type) {
+  return (has_nulls ? "optional " : "") + type;
+}
+
+// Inferred JSON types for fields.
+// Internally, the type is maintained along with the hitsogram of its values seen.
+CURRENT_STRUCT(String) {
+  CURRENT_FIELD(values, (std::map<std::string, size_t>));
+  CURRENT_FIELD(instances, size_t, 1);
+  CURRENT_FIELD(nulls, size_t, 0);
```
Comment (mzhurovich) [inline Utils/JSONSchema/infer.h]: `size_t` is platform-dependent and should not be used in `CURRENT_STRUCT`.
In particular, on my machine it becomes `unsigned long`, but there's no template specialization for `unsigned long` in our type system => it doesn't compile.

File: Utils/JSONSchema/infer.h [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+#include "../../TypeSystem/struct.h"
+#include "../../TypeSystem/Schema/schema.h"
+#include "../../TypeSystem/Serialization/json.h"
+
+namespace current {
+namespace utils {
+
+struct InferSchemaException : Exception {
+  InferSchemaException() : Exception() {}
+  explicit InferSchemaException(const std::string& message) : Exception(message) {}
```
Comment (mzhurovich) [inline Utils/JSONSchema/infer.h]: I see you don't like `using` constructors :)

File: Utils/JSONSchema/infer.h [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+
+struct InferSchemaArrayAndObjectAreIncompatible : InferSchemaIncompatibleTypesBase {};
+
+namespace impl {
+
+inline bool IsValidCPPIdentifier(const std::string& s) {
+  if (s.empty()) {
+    return false;
+  }
+  if (s == "NULL" || s == "null" || s == "true" || s == "false") {
```
Comment (mzhurovich) [inline Utils/JSONSchema/infer.h]: `const std::set<std::string>` with invalid identifiers?

File: Utils/JSONSchema/infer.h [hunk 4]
```diff
@@ -0,0 +1,542 @@
...
+  }
+  return true;
+}
+
+inline std::string MaybeOptionalHumanReadableType(bool has_nulls, const std::string& type) {
+  return (has_nulls ? "optional " : "") + type;
+}
+
+// Inferred JSON types for fields.
+// Internally, the type is maintained along with the hitsogram of its values seen.
```
Comment (mzhurovich) [inline Utils/JSONSchema/infer.h]: s/hitsogram/histogram/

File: Utils/JSONSchema/infer.h [hunk 5]
```diff
@@ -0,0 +1,542 @@
...
+
+    void operator()(const String& x) { output_type = x.nulls ? "Optional<std::string>" : "std::string"; }
+
+    void operator()(const Bool& x) { output_type = x.nulls ? "Optional<bool>" : "bool"; }
+
+    void operator()(const ObjectOrArray& x) {
+      if (x.IsObject()) {
+        output_type = prefix + "_Object";
+        std::string type;
+        std::string comment;
```
Comment (mzhurovich) [inline Utils/JSONSchema/infer.h]: Looks like these two strings are not used below.

File: Utils/JSONSchema/infer.h (hunk 3 above)
Comment (dkorolev) [inline Utils/JSONSchema/infer.h]: Let's do it later, perhaps on a higher level.

File: Utils/JSONSchema/infer.h (hunk 2 above)
Comment (dkorolev) [inline Utils/JSONSchema/infer.h]: I do :+1: 4bc59f4

File: Utils/JSONSchema/infer.h (hunk 1 above)
Comment (dkorolev) [inline Utils/JSONSchema/infer.h]: Good catch, my bad. 4bc59f4

And added a TODO to add this check into `CURRENT_FIELD` definition.

File: Utils/JSONSchema/infer.h (hunk 4 above)
Comment (dkorolev) [inline Utils/JSONSchema/infer.h]: :+1: 4bc59f4

File: Utils/JSONSchema/infer.h (hunk 5 above)
Comment (dkorolev) [inline Utils/JSONSchema/infer.h]: :+1: 4bc59f4

**Comment by dkorolev** [comment]
@mzhurovich Thanks for the review! PTAL.

**Comment by mzhurovich** [comment]
@dkorolev Looks good! Merging.

**Pull request merged**